src/typecheck/trans_sys.ml

```diff
@@ -0,0 +1,246 @@
+
+  let parse_sig env input output local =
+    let envs = env, env in
+    let envs, parsed_input  = parse_def_params envs input in
+    let envs, parsed_output = parse_def_params envs output in
+    let envs, parsed_local  = parse_def_params envs local in
+    envs, parsed_input, parsed_output, parsed_local
+
+  let parse_condition env cond =
+    Type.parse_prop env cond |> ignore
+
+  let _cannot_find_system env loc id =
+    Type._error env (Located loc) (Cannot_find_system id)
+
+  let _bad_inst_arity env loc id e a =
+    Type._error env (Located loc) (Bad_inst_arity (id, e, a))
+
+  let _duplicate_definition env loc1 id loc2 =
+    let loc1, loc2 =
+      if Loc.compare loc1 loc2 < 0 then loc2, loc1 else loc1, loc2
+    in
+    Type._error env (Located loc1) (Duplicate_definition (id, loc2))
+
+  let ensure env ast t ty =
+    Type._wrap2 env ast Type.T.ensure t ty
+
+  let parse_ensure env ast ty =
+    let t = Type.parse_term env ast in
+    ensure env ast t ty
+
+  let vars l = List.map (fun (_, v, _) -> v) l
+
+  let _bad_arity env s n m t =
+    Type._error env (Ast t) (Type.Bad_op_arity (s, [n], m))
+
+  let parse_subsystems env (parent : Stmt.sys_def) =
+    let defs = get_defs env in
+    List.fold_left
+      (fun other_subs (local_name, sub_inst, loc) ->
+        (* Make sure local name isn't used twice *)
+        match M.find_opt local_name other_subs with
+        | Some other_loc ->
+          _duplicate_definition env loc local_name other_loc
+        | None -> (
+          let sub_id, sid_loc, args, inst_loc = get_app_info sub_inst in
+          let sub_inputs, sub_outputs =
+            match M.find_opt sub_id defs with
+            | None ->
+                _cannot_find_system env sid_loc sub_id
+            | Some (`Trans_Sys (input, output, _)) ->
+                (vars input, vars output)
+          in
+          let num_args = List.length args in
+          let params = sub_inputs @ sub_outputs in
+          let num_params = List.length params in
+          if (num_args != num_params) then (
+            _bad_inst_arity env inst_loc sub_id num_params num_args
+          ) ;
+          List.iter2
+            (fun arg param ->
+              let expected_type = Type.T.Var.ty param in
+              parse_ensure env arg expected_type |> ignore
+            )
+            args
+            params ;
+          M.add local_name loc other_subs
+        )
+      )
+      M.empty
+      parent.subs
+    |> ignore
+
+  let parse_def_body ((env, env'), _input, _output, _local) (d: Stmt.sys_def) =
+    parse_condition env d.init ;
+    parse_condition env' d.trans ; 
+    parse_condition env d.inv ;
+    parse_subsystems env d
+
+  let finalize_sys (d : Stmt.sys_def) ((env, _), input, output, local) =
+    Type.check_no_free_wildcards env d.init;
+    Type.check_no_free_wildcards env d.trans;
+    Type.check_no_free_wildcards env d.inv;
+    let input, output, local = vars input, vars output, vars local in
```

Comment:
This way of writing code is a bit mislead I think. Instead I think it'd be better to 'inline' the call to `vars` for each one of the `List.iter` below.